In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/weather-data/Weather Data.csv


### IMPORT MODULES AND DATASET

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv("../input/weather-data/Weather Data.csv")

In [4]:
data

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

### DATA EXPLORATION

Before Analyze data : we will explore data first!!
.head( )
.shape
.index
.column



In [5]:
data.head

<bound method NDFrame.head of              Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Vis

In [6]:
data.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

In [7]:
data.shape

(8784, 8)

total columns and rows

In [8]:
data.index


RangeIndex(start=0, stop=8784, step=1)

In [9]:
data.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

name of each column

In [10]:
data.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

Show the unique value of the selected single column

In [11]:
data['Press_kPa'].unique()

array([101.24, 101.26, 101.27, 101.23, 101.29, 101.2 , 101.15, 100.98,
       100.79, 100.58, 100.31, 100.07,  99.93,  99.81,  99.74,  99.68,
        99.5 ,  99.39,  99.32,  99.31,  99.26,  99.3 ,  99.21,  99.18,
        99.14,  99.19,  99.27,  99.33,  99.41,  99.66,  99.86, 100.  ,
       100.15, 100.26, 100.27, 100.35, 100.43, 100.53, 100.61, 100.68,
       100.76, 100.85, 100.95, 101.07, 101.33, 101.34, 101.38, 101.44,
       101.51, 101.58, 101.64, 101.74, 101.82, 101.86, 101.87, 101.88,
       101.89, 101.77, 101.75, 101.73, 101.68, 101.54, 101.41, 101.25,
       100.97, 100.89, 100.69, 100.65, 100.54, 100.5 , 100.42, 100.32,
       100.29, 100.28, 100.24, 100.21, 100.19, 100.39, 100.45, 100.41,
       100.38, 100.4 , 100.46, 100.52, 100.6 , 100.71, 100.8 , 100.83,
       100.81, 100.84, 100.7 , 100.51, 100.47, 100.36, 100.23, 100.13,
       100.06, 100.1 , 100.12, 100.16, 100.11, 100.08, 100.05, 100.09,
       100.33, 100.48, 100.62, 100.72, 101.13, 101.21, 101.47, 101.59,
      

nunique() : can apply on single column or whole dataframes

In [12]:
data['Press_kPa'].nunique()

518

In [13]:
data.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

example : weather column has 518 unique weather condition values

In [14]:
data.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

In [15]:
data['Weather'].value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

example: freezing rain = 14 time-counts

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


shows the dataset info

In [17]:
data.head(2)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   

   Visibility_km  Press_kPa Weather  
0            8.0     101.24     Fog  
1            8.0     101.24     Fog

In [18]:
data['Visibility_km'].nunique()

24

--unique count :24,
--and give me the number of count for each unique value,
--like value 4.0 : count 55 times

In [19]:
data['Visibility_km'].value_counts()

25.0    3324
48.3    2014
24.1    1921
19.3     281
16.1     224
9.7      181
12.9     174
6.4      129
8.0      122
4.8       79
4.0       55
2.4       50
3.2       42
11.3      36
1.2       28
2.0       28
1.6       25
0.6       16
2.8       15
1.0       11
3.6       11
0.2        8
0.8        7
0.4        3
Name: Visibility_km, dtype: int64

In [20]:
data['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

show me : unique value of wind speed


and (below) show me number of those value

In [21]:
data['Wind Speed_km/h'].nunique()

34

## Filtering

In [22]:
data.Weather.value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

would like to get only the days which have clear weather

In [23]:
### data[dataframe]

data[data.Weather == 'Clear']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

1326 days

### GROUP BY

In [24]:
data.groupby('Weather').get_group('Clear')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
114     1/5/2012 18:00    -7.1             -14.4         56               11   
115     1/5/2012 19:00    -9.2             -15.4         61                7   
116     1/5/2012 20:00    -9.8             -15.7         62                9   
117     1/5/2012 21:00    -9.0             -14.8         63               13   
...                ...     ...               ...        ...              ...   
8646   12/26/2012 6:00   -13.4             -14.8         89                4   
8698  12/28/2012 10:00    -6.1              -8.6         82               19   
8713   12/29/2012 1:00   -11.9             -13.6         87               11   
8714   12/29/2012 2:00   -11.8             -13.1         90               13   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather  
67             25.0     101.74   Clear  
114            25.0     100.71   Clear  
115            25.0     100.80   Clear  
116            25.0     100.83   Clear  
117            25.0     100.83   Clear  
...             ...        ...     ...  
8646           25.0     102.47   Clear  
8698           24.1     101.27   Clear  
8713           25.0     101.31   Clear  
8714           25.0     101.33   Clear  
8756           25.0     101.52   Clear  

[1326 rows x 8 columns]

In [25]:
## Wind speed = 4km/h

data.groupby('Wind Speed_km/h').get_group(4)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

474 days for wind speed = 4km/h

In [26]:
data[data['Wind Speed_km/h'] == 4]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
96       1/5/2012 0:00    -8.8             -11.7         79                4   
101      1/5/2012 5:00    -7.0              -9.5         82                4   
146      1/7/2012 2:00    -8.1             -11.1         79                4   
...                ...     ...               ...        ...              ...   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   

      Visibility_km  Press_kPa       Weather  
0               8.0     101.24           Fog  
1               8.0     101.24           Fog  
96              9.7     100.32          Snow  
101             4.0     100.19          Snow  
146            19.3     100.15        Cloudy  
...             ...        ...           ...  
8768            3.2     101.14  Snow Showers  
8769            2.4     101.09          Snow  
8770            6.4     101.05      Snow,Fog  
8772           12.9     100.78          Snow  
8773           12.9     100.63          Snow  

[474 rows x 8 columns]

### Finding NULL

In [27]:
data.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

In [28]:
data.notnull().sum()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

### RENAMING

In [29]:
data.rename(columns ={'Weather' : 'Weather Condition'})

### but rename only this command

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa     Weather Condition  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

In [30]:
## if rename permanently

data.rename(columns = {'Weather' : 'Weather Condition'}, inplace = True)

In [31]:
data

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa     Weather Condition  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

### MEAN

In [32]:
data.Visibility_km.mean()

27.66444672131151

### STANDARD DEVIATION

In [33]:
data.Press_kPa.std()

0.8440047459486474

### Variance of 

In [34]:
data['Rel Hum_%'].var()

286.2485501984998

NOTE *** coumn name which had space in name ,we will use[], if no space, we can use dot(.)

In [35]:
data.groupby('Weather Condition').get_group('Snow')

## value_counts ::: value "Weather Condition" -- count "Snow" for me

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
55       1/3/2012 7:00   -14.0             -19.5         63               19   
84      1/4/2012 12:00   -13.7             -21.7         51               11   
86      1/4/2012 14:00   -11.3             -19.0         53                7   
87      1/4/2012 15:00   -10.2             -16.3         61               11   
88      1/4/2012 16:00    -9.4             -15.5         61               13   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
55             25.0     100.95              Snow  
84             24.1     101.25              Snow  
86             19.3     100.97              Snow  
87              9.7     100.89              Snow  
88             19.3     100.79              Snow  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[390 rows x 8 columns]

In [36]:
## or

data['Weather Condition'].value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

Snow = 390

In [37]:
## or filtering
data[data['Weather Condition'] == 'Snow']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
55       1/3/2012 7:00   -14.0             -19.5         63               19   
84      1/4/2012 12:00   -13.7             -21.7         51               11   
86      1/4/2012 14:00   -11.3             -19.0         53                7   
87      1/4/2012 15:00   -10.2             -16.3         61               11   
88      1/4/2012 16:00    -9.4             -15.5         61               13   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
55             25.0     100.95              Snow  
84             24.1     101.25              Snow  
86             19.3     100.97              Snow  
87              9.7     100.89              Snow  
88             19.3     100.79              Snow  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[390 rows x 8 columns]

 but they show many type of snow days :- shower snow, snow fog , snow haze, snow blowing, snow rain 

### str.contains

In [38]:
data[data['Weather Condition'].str.contains('Snow')]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
41      1/2/2012 17:00    -2.1              -9.5         57               22   
44      1/2/2012 20:00    -5.6             -13.4         54               24   
45      1/2/2012 21:00    -5.8             -12.8         58               26   
47      1/2/2012 23:00    -7.4             -14.1         59               17   
48       1/3/2012 0:00    -9.0             -16.0         57               28   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather Condition  
41             25.0      99.66      Snow Showers  
44             25.0     100.07      Snow Showers  
45             25.0     100.15      Snow Showers  
47             19.3     100.27      Snow Showers  
48             25.0     100.35      Snow Showers  
...             ...        ...               ...  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow  

[583 rows x 8 columns]

In [39]:
data[data['Weather Condition'].str.contains('Snow')].tail(50)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
8680  12/27/2012 16:00    -4.5              -6.2         88               37   
8681  12/27/2012 17:00    -4.2              -5.9         88               32   
8682  12/27/2012 18:00    -4.0              -5.7         88               28   
8683  12/27/2012 19:00    -3.9              -5.6         88               26   
8684  12/27/2012 20:00    -3.7              -5.3         89               37   
8685  12/27/2012 21:00    -3.7              -4.8         92               24   
8686  12/27/2012 22:00    -3.8              -4.6         94               20   
8687  12/27/2012 23:00    -4.0              -5.6         89               24   
8688   12/28/2012 0:00    -4.2              -5.7         89               19   
8689   12/28/2012 1:00    -4.4              -6.6         85               15   
8690   12/28/2012 2:00    -4.3              -6.3         86               11   
8691   12/28/2012 3:00    -4.6              -5.9         91               13   
8692   12/28/2012 4:00    -4.9              -5.9         93                9   
8723  12/29/2012 11:00   -10.9             -12.2         90                7   
8724  12/29/2012 12:00   -10.5             -11.6         92               11   
8725  12/29/2012 13:00   -10.0             -11.1         92               22   
8726  12/29/2012 14:00    -9.3             -10.5         91               22   
8727  12/29/2012 15:00    -8.8             -10.0         91               20   
8728  12/29/2012 16:00    -8.5              -9.9         90               24   
8729  12/29/2012 17:00    -9.0             -10.4         90               19   
8730  12/29/2012 18:00    -9.3             -10.9         88               26   
8731  12/29/2012 19:00    -9.5             -11.2         87               26   
8732  12/29/2012 20:00    -9.7             -11.6         86               24   
8733  12/29/2012 21:00    -9.8             -11.8         85               24   
8734  12/29/2012 22:00   -10.1             -11.6         89               15   
8735  12/29/2012 23:00   -10.0             -12.0         85               20   
8736   12/30/2012 0:00    -9.6             -11.3         87               13   
8737   12/30/2012 1:00    -9.4             -10.5         92                9   
8738   12/30/2012 2:00    -9.3             -10.4         92                9   
8739   12/30/2012 3:00    -9.1             -10.4         90               11   
8740   12/30/2012 4:00    -9.3             -10.6         90               13   
8741   12/30/2012 5:00    -9.1             -10.4         90               11   
8742   12/30/2012 6:00    -9.3             -10.8         89               17   
8767   12/31/2012 7:00    -9.3             -11.3         85                0   
8768   12/31/2012 8:00    -8.6             -10.3         87                4   
8769   12/31/2012 9:00    -8.1              -9.6         89                4   
8770  12/31/2012 10:00    -7.4              -8.9         89                4   
8771  12/31/2012 11:00    -6.7              -7.9         91                9   
8772  12/31/2012 12:00    -5.8              -7.5         88                4   
8773  12/31/2012 13:00    -4.6              -6.6         86                4   
8774  12/31/2012 14:00    -3.4              -5.7         84                6   
8775  12/31/2012 15:00    -2.3              -4.6         84                9   
8776  12/31/2012 16:00    -1.4              -4.0         82               13   
8777  12/31/2012 17:00    -1.1              -3.3         85               19   
8778  12/31/2012 18:00    -1.3              -3.1         88               17   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2

pure snow ; 390 days
but snow other kinds : 583 days total

### AND

In [40]:
## want to get 'wind speed >24km/h' and 'visibility'=25km

data[(data['Wind Speed_km/h'] > 24) & (data['Visibility_km'] == 25)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
23      1/1/2012 23:00     5.3               2.0         79               30   
24       1/2/2012 0:00     5.2               1.5         77               35   
25       1/2/2012 1:00     4.6               0.0         72               39   
26       1/2/2012 2:00     3.9              -0.9         71               32   
27       1/2/2012 3:00     3.7              -1.5         69               33   
...                ...     ...               ...        ...              ...   
8705  12/28/2012 17:00    -8.6             -12.0         76               26   
8753  12/30/2012 17:00   -12.1             -15.8         74               28   
8755  12/30/2012 19:00   -13.4             -16.5         77               26   
8759  12/30/2012 23:00   -12.1             -15.1         78               28   
8760   12/31/2012 0:00   -11.1             -14.4         77               26   

      Visibility_km  Press_kPa Weather Condition  
23             25.0      99.31            Cloudy  
24             25.0      99.26      Rain Showers  
25             25.0      99.26            Cloudy  
26             25.0      99.26     Mostly Cloudy  
27             25.0      99.30     Mostly Cloudy  
...             ...        ...               ...  
8705           25.0     101.34      Mainly Clear  
8753           25.0     101.26      Mainly Clear  
8755           25.0     101.47      Mainly Clear  
8759           25.0     101.52     Mostly Cloudy  
8760           25.0     101.51            Cloudy  

[308 rows x 8 columns]

In [41]:
## mean of each column  against each weather condition

data.groupby('Weather Condition').mean()

Temp_C  Dew Point Temp_C  \
Weather Condition                                                      
Clear                                     6.825716          0.089367   
Cloudy                                    7.970544          2.375810   
Drizzle                                   7.353659          5.504878   
Drizzle,Fog                               8.067500          7.033750   
Drizzle,Ice Pellets,Fog                   0.400000         -0.700000   
Drizzle,Snow                              1.050000          0.150000   
Drizzle,Snow,Fog                          0.693333          0.120000   
Fog                                       4.303333          3.159333   
Freezing Drizzle                         -5.657143         -8.000000   
Freezing Drizzle,Fog                     -2.533333         -4.183333   
Freezing Drizzle,Haze                    -5.433333         -8.000000   
Freezing Drizzle,Snow                    -5.109091         -7.072727   
Freezing Fog                             -7.575000         -9.250000   
Freezing Rain                            -3.885714         -6.078571   
Freezing Rain,Fog                        -2.225000         -3.750000   
Freezing Rain,Haze                       -4.900000         -7.450000   
Freezing Rain,Ice Pellets,Fog            -2.600000         -3.700000   
Freezing Rain,Snow Grains                -5.000000         -7.300000   
Haze                                     -0.200000         -2.975000   
Mainly Clear                             12.558927          4.581671   
Moderate Rain,Fog                         1.700000          0.800000   
Moderate Snow                            -5.525000         -7.250000   
Moderate Snow,Blowing Snow               -5.450000         -6.500000   
Mostly Cloudy                            10.574287          3.131174   
Rain                                      9.786275          7.042810   
Rain Showers                             13.722340          9.187766   
Rain Showers,Fog                         12.800000         12.100000   
Rain Showers,Snow Showers                 2.150000         -1.500000   
Rain,Fog                                  8.273276          7.219828   
Rain,Haze                                 4.633333          2.066667   
Rain,Ice Pellets                          0.600000         -0.600000   
Rain,Snow                                 1.055556         -0.566667   
Rain,Snow Grains                          1.900000         -2.100000   
Rain,Snow,Fog                             0.800000          0.300000   
Rain,Snow,Ice Pellets                     1.100000         -0.175000   
Snow                                     -4.524103         -7.623333   
Snow Pellets                              0.700000         -6.400000   
Snow Showers                             -3.506667         -7.866667   
Snow Showers,Fog                        -10.675000        -11.900000   
Snow,Blowing Snow                        -5.410526         -7.621053   
Snow,Fog                                 -5.075676         -6.364865   
Snow,Haze                                -4.020000         -6.860000   
Snow,Ice Pellets                         -1.883333         -3.666667   
Thunderstorms                            24.150000         19.750000   
Thunderstorms,Heavy Rain Showers         10.900000          9.000000   
Thunderstorms,Moderate Rain Showers,Fog  19.600000         18.500000   
Thunderstorms,Rain                       20.433333         18.533333   
Thunderstorms,Rain Showers               20.037500         17.618750   
Thunderstorms,Rain Showers,Fog           21.600000         18.700000   
Thunderstorms,Rain,Fog                   20.600000         18.600000   

                                         Rel Hum_%  Wind Speed_km/h  \
Weather Condition                                                     
Clear                                    64.497738        10.557315   
Cloudy                                   69.592593        16.127315   
Drizzle        

### MIN , MAX

In [42]:
### min , max for each column against each weather condition

data.groupby('Weather Condition').min()

Date/Time  Temp_C  \
Weather Condition                                                   
Clear                                      1/11/2012 1:00   -23.3   
Cloudy                                     1/1/2012 17:00   -21.4   
Drizzle                                   1/23/2012 21:00     1.1   
Drizzle,Fog                               1/23/2012 20:00     0.0   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/17/2012 15:00     0.9   
Drizzle,Snow,Fog                         12/18/2012 21:00     0.3   
Fog                                         1/1/2012 0:00   -16.0   
Freezing Drizzle                          1/13/2012 10:00    -9.0   
Freezing Drizzle,Fog                        1/1/2012 2:00    -6.4   
Freezing Drizzle,Haze                      2/1/2012 11:00    -5.8   
Freezing Drizzle,Snow                      1/13/2012 3:00    -8.3   
Freezing Fog                               1/22/2012 6:00   -19.0   
Freezing Rain                             1/13/2012 11:00    -6.5   
Freezing Rain,Fog                         1/17/2012 23:00    -6.1   
Freezing Rain,Haze                         2/1/2012 14:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      1/22/2012 12:00   -11.5   
Mainly Clear                              1/10/2012 11:00   -22.8   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             1/12/2012 15:00    -6.3   
Moderate Snow,Blowing Snow               12/27/2012 10:00    -5.5   
Mostly Cloudy                              1/1/2012 16:00   -23.2   
Rain                                       1/1/2012 18:00     0.3   
Rain Showers                               1/1/2012 22:00     1.6   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                  11/4/2012 8:00     2.1   
Rain,Fog                                  1/23/2012 18:00     0.0   
Rain,Haze                                  3/13/2012 7:00     4.0   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  1/10/2012 5:00     0.6   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 1:00     0.9   
Snow                                       1/10/2012 1:00   -16.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               1/12/2012 7:00   -13.3   
Snow Showers,Fog                          12/26/2012 9:00   -11.3   
Snow,Blowing Snow                         1/13/2012 21:00   -12.0   
Snow,Fog                                 12/16/2012 15:00   -10.1   
Snow,Haze                                  2/1/2012 17:00    -4.3   
Snow,Ice Pellets                          12/10/2012 3:00    -4.3   
Thunderstorms                              7/16/2012 1:00    21.6   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        5/25/2012 20:00    19.4   
Thunderstorms,Rain Showers                5/29/2012 16:00    11.0   
Thunderstorms,Rain Showers,Fog             6/29/2012 3:00    19.5   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                               -28.5         20   
Cloudy                                              -26.8         18   
Drizzle                                              -0.2         74   
Drizzle,Fog                                          -1.6         85   
Drizzle,Ice Pellets,Fog    

In [43]:
data.groupby('Weather Condition').max()

Date/Time  Temp_C  \
Weather Condition                                                   
Clear                                       9/9/2012 5:00    32.8   
Cloudy                                     9/9/2012 23:00    30.5   
Drizzle                                    9/30/2012 3:00    18.8   
Drizzle,Fog                                9/30/2012 2:00    19.9   
Drizzle,Ice Pellets,Fog                   12/17/2012 9:00     0.4   
Drizzle,Snow                             12/19/2012 18:00     1.2   
Drizzle,Snow,Fog                          12/22/2012 3:00     1.1   
Fog                                        9/22/2012 0:00    20.8   
Freezing Drizzle                            2/1/2012 5:00    -2.3   
Freezing Drizzle,Fog                      12/10/2012 5:00    -0.3   
Freezing Drizzle,Haze                      2/1/2012 13:00    -5.0   
Freezing Drizzle,Snow                      3/2/2012 12:00    -3.3   
Freezing Fog                               3/17/2012 6:00    -0.1   
Freezing Rain                               2/1/2012 7:00     0.3   
Freezing Rain,Fog                         12/17/2012 1:00     0.1   
Freezing Rain,Haze                         2/1/2012 15:00    -4.9   
Freezing Rain,Ice Pellets,Fog             12/17/2012 3:00    -2.6   
Freezing Rain,Snow Grains                  1/13/2012 9:00    -5.0   
Haze                                      3/13/2012 23:00    14.1   
Mainly Clear                                9/9/2012 9:00    33.0   
Moderate Rain,Fog                         12/10/2012 8:00     1.7   
Moderate Snow                             12/27/2012 9:00    -4.9   
Moderate Snow,Blowing Snow               12/27/2012 12:00    -5.4   
Mostly Cloudy                               9/9/2012 2:00    32.4   
Rain                                        9/5/2012 2:00    22.8   
Rain Showers                               9/8/2012 16:00    26.4   
Rain Showers,Fog                          10/20/2012 3:00    12.8   
Rain Showers,Snow Showers                 12/5/2012 10:00     2.2   
Rain,Fog                                  9/30/2012 23:00    21.7   
Rain,Haze                                  3/13/2012 9:00     5.5   
Rain,Ice Pellets                          12/18/2012 5:00     0.6   
Rain,Snow                                  4/23/2012 3:00     1.7   
Rain,Snow Grains                          12/21/2012 0:00     1.9   
Rain,Snow,Fog                             12/8/2012 21:00     0.8   
Rain,Snow,Ice Pellets                     12/21/2012 5:00     1.3   
Snow                                       4/27/2012 9:00     3.7   
Snow Pellets                             11/24/2012 15:00     0.7   
Snow Showers                               3/4/2012 21:00     2.9   
Snow Showers,Fog                         12/29/2012 13:00   -10.0   
Snow,Blowing Snow                          2/25/2012 9:00    -1.4   
Snow,Fog                                  3/14/2012 19:00     1.1   
Snow,Haze                                  2/1/2012 21:00    -3.6   
Snow,Ice Pellets                            3/3/2012 4:00     0.8   
Thunderstorms                              7/4/2012 16:00    26.7   
Thunderstorms,Heavy Rain Showers           5/29/2012 6:00    10.9   
Thunderstorms,Moderate Rain Showers,Fog    7/17/2012 6:00    19.6   
Thunderstorms,Rain                        7/23/2012 18:00    21.3   
Thunderstorms,Rain Showers                  9/8/2012 4:00    25.5   
Thunderstorms,Rain Showers,Fog            7/31/2012 20:00    22.9   
Thunderstorms,Rain,Fog                     7/17/2012 5:00    20.6   

                                         Dew Point Temp_C  Rel Hum_%  \
Weather Condition                                                      
Clear                                                20.4         99   
Cloudy                                               22.6         99   
Drizzle                                              17.7         96   
Drizzle,Fog                                          19.1        100   
Drizzle,Ice Pellets,Fog    

In [44]:
## show all records of "Foggy"day
data[data['Weather Condition'] == 'Fog']

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather Condition  
0               8.0     101.24               Fog  
1               8.0     101.24               Fog  
4               4.8     101.23               Fog  
5               6.4     101.27               Fog  
6               6.4     101.29               Fog  
...             ...        ...               ...  
8716            9.7     101.25               Fog  
8717            6.4     101.25               Fog  
8718            9.7     101.25               Fog  
8719            8.0     101.22               Fog  
8722            6.4     101.15               Fog  

[150 rows x 8 columns]

In [45]:
data.groupby('Weather Condition').get_group('Fog')

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
5        1/1/2012 5:00    -1.4              -3.3         87                9   
6        1/1/2012 6:00    -1.5              -3.1         89                7   
...                ...     ...               ...        ...              ...   
8716   12/29/2012 4:00   -16.0             -17.2         90                6   
8717   12/29/2012 5:00   -14.8             -15.9         91                4   
8718   12/29/2012 6:00   -13.8             -15.3         88                4   
8719   12/29/2012 7:00   -14.8             -16.4         88                7   
8722  12/29/2012 10:00   -12.0             -13.3         90                7   

      Visibility_km  Press_kPa Weather Condition  
0               8.0     101.24               Fog  
1               8.0     101.24               Fog  
4               4.8     101.23               Fog  
5               6.4     101.27               Fog  
6               6.4     101.29               Fog  
...             ...        ...               ...  
8716            9.7     101.25               Fog  
8717            6.4     101.25               Fog  
8718            9.7     101.25               Fog  
8719            8.0     101.22               Fog  
8722            6.4     101.15               Fog  

[150 rows x 8 columns]

### OR

In [46]:
## clear weather and visibility >40

data[(data['Weather Condition'] == 'Clear') | (data['Visibility_km'] > 40)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67      1/3/2012 19:00   -16.9             -24.8         50               24   
106     1/5/2012 10:00    -6.0             -10.0         73               17   
107     1/5/2012 11:00    -5.6             -10.2         70               22   
108     1/5/2012 12:00    -4.7              -9.6         69               20   
109     1/5/2012 13:00    -4.4              -9.7         66               26   
...                ...     ...               ...        ...              ...   
8749  12/30/2012 13:00   -12.4             -16.2         73               37   
8750  12/30/2012 14:00   -11.8             -16.1         70               37   
8751  12/30/2012 15:00   -11.3             -15.6         70               32   
8752  12/30/2012 16:00   -11.4             -15.5         72               26   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather Condition  
67             25.0     101.74             Clear  
106            48.3     100.45      Mainly Clear  
107            48.3     100.41      Mainly Clear  
108            48.3     100.38      Mainly Clear  
109            48.3     100.40      Mainly Clear  
...             ...        ...               ...  
8749           48.3     100.92     Mostly Cloudy  
8750           48.3     100.96      Mainly Clear  
8751           48.3     101.05      Mainly Clear  
8752           48.3     101.15      Mainly Clear  
8756           25.0     101.52             Clear  

[3027 rows x 8 columns]

In [47]:
## weather clear (A) and relative humidity more than 50 (B) or visibity >40 (C)

## (A and B) or C

data[(data['Weather Condition'] == 'Clear') & (data['Rel Hum_%'] > 50) | (data['Visibility_km'] > 40)]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
106     1/5/2012 10:00    -6.0             -10.0         73               17   
107     1/5/2012 11:00    -5.6             -10.2         70               22   
108     1/5/2012 12:00    -4.7              -9.6         69               20   
109     1/5/2012 13:00    -4.4              -9.7         66               26   
110     1/5/2012 14:00    -5.1             -10.7         65               22   
...                ...     ...               ...        ...              ...   
8749  12/30/2012 13:00   -12.4             -16.2         73               37   
8750  12/30/2012 14:00   -11.8             -16.1         70               37   
8751  12/30/2012 15:00   -11.3             -15.6         70               32   
8752  12/30/2012 16:00   -11.4             -15.5         72               26   
8756  12/30/2012 20:00   -13.8             -16.5         80               24   

      Visibility_km  Press_kPa Weather Condition  
106            48.3     100.45      Mainly Clear  
107            48.3     100.41      Mainly Clear  
108            48.3     100.38      Mainly Clear  
109            48.3     100.40      Mainly Clear  
110            48.3     100.46      Mainly Clear  
...             ...        ...               ...  
8749           48.3     100.92     Mostly Cloudy  
8750           48.3     100.96      Mainly Clear  
8751           48.3     101.05      Mainly Clear  
8752           48.3     101.15      Mainly Clear  
8756           25.0     101.52             Clear  

[2921 rows x 8 columns]